network/framework/src/protocols/health_checker/mod.rs

```diff
@@ -143,6 +155,13 @@ impl<NetworkClient: NetworkClientInterface<HealthCheckerMsg> + Unpin> HealthChec
         let ticker = self.time_service.interval(self.ping_interval);
         tokio::pin!(ticker);
 
+        let connection_events = self
```

Comment:
Is it possible for connections to already exist when we get here? If so, `subscribe()` will miss them, right?

Reply:
Maybe, unlikely, and no greater chance of race than current code.

Reply:
> and no greater chance of race than current code.

Why not make things better along the way? 😄

Reply:
Can we ensure that all consumers of NewPeer are idempotent and that a NewPeer event for something already known is okay? If so then sure, `subscribe()` could start by queuing up a NewPeer event for everything currently connected.

Reply:
> Can we ensure that all consumers of NewPeer are idempotent and that a NewPeer event for something already known is okay?

Why is this the case? `subscribe()` should grab the write lock on the peers and metadata struct, immediately fire events for all active peers, and then return. If a peer is added/removed during this time, it'll block until `subscribe()` is finished (because it also needs the peers and metadata lock) and then the correct event will be queued up? 😄

Reply:
subscribe() implemented as described!